File: src/games_sims3/sims3_game_profile.cpp
#include "sxpe/games/sims3/sims3_game_profile.hpp"

#include <cstdint>
#include <cstring>

namespace sxpe::games::sims3 {

Sims3GameProfile::Sims3GameProfile() : codec_(*this) {}

std::span<const std::string_view> Sims3GameProfile::file_extensions() const {
    return kExts;
}

bool Sims3GameProfile::try_sniff(std::span<const std::byte> header, float& confidence) const {
    confidence = 0.f;
    if (header.size() < 4) {
        return false;
    }
    if (header[0] != std::byte{'D'} || header[1] != std::byte{'B'} || header[2] != std::byte{'P'} ||
        header[3] != std::byte{'F'}) {
        return false;
    }
    if (header.size() >= 8) {
        std::uint32_t major = 0;
        std::memcpy(&major, header.data() + 4, 4);
        if (major != 2) {
            return false;
        }
        confidence = 1.f;
        return true;
    }
    confidence = 0.5f;
    return true;
}

}  // namespace sxpe::games::sims3

File: src/Sxpe.Games.Sims3/Sims3GameProfile.cs
using Sxpe.Games;

namespace Sxpe.Games.Sims3;

public sealed class Sims3GameProfile : IGameProfile
{
    public Sims3GameProfile()
    {
        PackageCodec = new Sims3PackageCodec(this);
        Compression = new Sims3Compression();
    }

    public GameId Id => GameId.Sims3;

    public string DisplayName => "The Sims 3";

    public IReadOnlyList<string> FileExtensions { get; } =
        [".package", ".world", ".dbc", ".nhd"];

    public IPackageCodec PackageCodec { get; }

    public ICompressionCodec Compression { get; }

    /// <summary>
    /// Header sniff is filled in when the DBPF reader lands.
    /// Until then this profile never claims a file.
    /// </summary>
    public bool TrySniff(ReadOnlySpan<byte> header, out float confidence)
    {
        confidence = 0;
        return false;
    }
}
